osutil/env.go

```diff
@@ -65,51 +68,145 @@ func GetenvInt64(key string, dflt ...int64) int64 {
+// key=value or if there are duplicate keys.
+func parseRawEnvironment(raw []string) (Environment, error) {
+	env := make(Environment, len(raw))
+	for _, entry := range raw {
+		key, value, err := parseEnvEntry(entry)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := env[key]; ok {
+			return nil, fmt.Errorf("cannot overwrite earlier value of %q", key)
 		}
+		env[key] = value
 	}
-	return out
+	return env, nil
 }
 
-// SubstituteEnv takes a list of environment strings like:
-// - K1=BAR
-// - K2=$K1
-// - K3=${K2}
-// and substitutes them top-down from the given environment
-// and from the os environment.
+// OSEnvironment returns the environment of the calling process.
+func OSEnvironment() (Environment, error) {
+	return parseRawEnvironment(os.Environ())
+}
+
+// ForExec returns environment suitable for using with exec family of functions.
 //
-// Input strings that do not have the form "k=v" will be dropped
-// from the output.
+// The returned environment is sorted lexicographically by variable name.
+func (env Environment) ForExec() []string {
+	raw := make([]string, 0, len(env))
+	keys := make([]string, 0, len(env))
+	for key := range env {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		raw = append(raw, fmt.Sprintf("%s=%s", key, env[key]))
+	}
+	return raw
+}
+
+// ExpandableEnv describes alterations to an environment.
 //
-// The result will be a list of environment strings in the same
-// order as the input.
-func SubstituteEnv(env []string) []string {
-	envMap := map[string]string{}
-	out := make([]string, 0, len(env))
-
-	for _, s := range env {
-		l := strings.SplitN(s, "=", 2)
-		if len(l) < 2 {
-			continue
-		}
-		k := l[0]
-		v := l[1]
-		v = os.Expand(v, func(k string) string {
-			if s, ok := envMap[k]; ok {
-				return s
-			}
-			return os.Getenv(k)
+// Differential environment can refer to existing entries by using shell-like
+// syntax $KEY or ${KEY}. Entries are ordered.
+type ExpandableEnv struct {
+	*strutil.OrderedMap
+}
+
+// NewExpandableEnv returns a new expandable environment comprised of given pairs.
+func NewExpandableEnv(pairs ...string) ExpandableEnv {
+	return ExpandableEnv{OrderedMap: strutil.NewOrderedMap(pairs...)}
+}
+
+// SetExpandableEnv sets variables defined by expandable environment
+//
+// Environment is modified in place. Each variable defined by expandable
+// environment is expanded according to os.Expand, using the environment itself
+// as data source. Undefined variables expand to an empty string.
+func (env *Environment) SetExpandableEnv(eenv ExpandableEnv) {
+	if *env == nil {
+		*env = make(Environment)
+	}
+
+	for _, key := range eenv.Keys() {
+		(*env)[key] = os.Expand(eenv.Get(key), func(varName string) string {
+			return (*env)[varName]
 		})
-		out = append(out, fmt.Sprintf("%s=%s", k, v))
-		envMap[k] = v
 	}
+}
 
-	return out
+func (env Environment) EscapeUnsafeVariables() {
```

Comment:
I would prefer this to be folded into a mode of ForExec via flags.

Reply:
We also lost some logic to drop already escaped variables that is on master, to be fair it was untested there, but probably best to preserve it.